src/stream/src/executor/merge.rs

```diff
@@ -464,6 +477,92 @@ impl SelectReceivers {
     }
 }
 
+/// A wrapper that buffers the `StreamChunk`s from upstream until no more ready items are available.
+/// Besides, any message other than `StreamChunk` will trigger the buffered `StreamChunk`s
+/// to be emitted immediately, as well as the message itself.
+struct BufferChunks<S: Stream> {
```

Comment:
Nit: Maybe we can call this `BufferedChunkReader` since it's very like `tokio::io::BufReader` and `java.io.BufferedReader`.